contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -761,3 +761,208 @@ The central manager will poll each worker for their discovered volumes and creat
+
+```
+ProvisionQueue *workqueue.Type
+```
+When it finds a PVC is in need of privsioning, as desicribed above,
+the `populator` will pass it into the queue to trigger volume privisioning;
+`StorageManager` will be the consumer, it will watch the queue and
+call its `CreateLocalVolume` function to:
+
+1. Create volumes behind a PV accordingly
+2. Create a PV object that is pre-bound to the PVC
+
+For LVM, the proposed workflow is:
+
+1. Check if there is already a PV pre-bound to the PVC, if so, return directly
+(This can happen when a volume has already been provisioned for a PVC,
+but has not been bound to the PVC by the controller yet).
+2. Check if the its capacity has been initialized in the StorageClass.
+If not, return error.
+3. Check if the class is available for dynamic provisioning (i.e. has an input of `volumeGroup`).
+If not, return error.
+4. Create `logical volume (LV)` object in the VG.
+5. Create a PV API object that is pre-bound to the PVC, and the path
+of local storage in the PV should be the path of the LV created above.
+
+**Note:** As it describes in [Volume Topology-aware Scheduling](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md),
+in whichever step it fails, the provisioner will signal back to the scheduler
+to retry dynamic provisioning by removing the "annSelectedNode" annotation of
+the PVC.
+
+For the purpose, it requires the permission to update PersistentVolumeClaims.
+
+##### Cleanup after Release
+
+Currently we already have a component `deleter` to cleanup volumes/PVs of static local volume,
+detail cleanup actions are managed via custom scripts/binaries that specified in `blockCleanerCommand`.
+
+Existing script `quick_reset.sh` calls `mkfs -F` and `wipefs` to cleanup a block.
+In addition to that, a hook named `volumeDeleteFunc` will be added into `deleter`,
+and for dynamic privisioned volumes, the deleting process should be:
+1. Call scripts/binaries in `blockCleanerCommand` to clean up data in the volume.
+2. Call `volumeDeleteFunc` to recycle the volume behind the PV (e.g. delete the LV for lvm).
+3. Delete the PV object.
+
+##### Expansion of Local Volume Plugin
+
+Currently, the path in `LocalVolumeSource` must be a directory. Thus, for dynamic
+provisioned volumes, the provisioner need to mount the LV to a host path first,
+to allow kubelet for another mount. It increases the complexity of the plugin.
+
+Therefore, in-tree local storage plugin will be updated to support both directory
+and block as a source of mount. The mount process would be:
+1. Detect if given path in volumeSource is mounted or not, if mounted,
+continue to work in the existing way.
+2. If not, Check if global path `{pluginDir}/mounts/{volumeID}` exists.
+and create the directory if not.
+3. Check if the global path is of mount point, if not, mount given path to
+the global path.
+4. Manage `mkfs` for the global path if it is newly created or mounted.

```

Comment:
> 4. Manage `mkfs` for the global path

You must call `mkfs` before mounting it. Internal volume plugins do it even before step 1.

Reply:
Yah, have not got a chance the update the section.
As discussed above, we need to consider this part along with the change in  #1683